src/js/child_process.js

```diff
@@ -567,81 +546,54 @@ function setupChannel(target, channel) {
 
     options = Object.assign({ swallowErrors: false }, options);
 
-    if (this.connected) {
-      return this._send(message, handle, options, callback);
-    }
-    var ex = new Error('ERR_IPC_CHANNEL_CLOSED');
-    if (typeof callback === 'function') {
-      process.nextTick(callback, ex);
-    } else {
-      process.nextTick(function() {
-        this.emit('error', ex);
-      }.bind(this));
+    if (!this.connected) {
+      var ex = new Error('ERR_IPC_CHANNEL_CLOSED');
+      if (typeof callback === 'function') {
+        process.nextTick(callback, ex);
+      } else {
+        process.nextTick(function() {
+          this.emit('error', ex);
+        }.bind(this));
+      }
+      return false;
     }
-    return false;
-  };
-
-  target._send = function(message, handle, options, callback) {
     if (message === undefined)
       throw new TypeError('ERR_MISSING_ARGS');
 
     // Support legacy function signature
     if (typeof options === 'boolean') {
       options = { swallowErrors: options };
     }
-
-    // package messages with a handle object
-    if (handle) {
-      // this message will be handled by an internalMessage event handler
-      message = {
-        cmd: 'NODE_HANDLE',
-        type: 'net.Native',
-        msg: message
-      };
-      // Queue-up message and handle if we haven't received ACK yet.
-      if (this._handleQueue) {
-        this._handleQueue.push({
-          callback: callback,
-          handle: handle,
-          options: options,
-          message: message.msg,
-        });
-        return this._handleQueue.length === 1;
+    var messageType;
+    if (typeof message === 'object') {
+      messageType = INTERNAL_IPC_MESSAGE_TYPE.OBJECT;
+      try {
+        message = JSON.stringify(message);
+      } catch (err) {
+        throw new Error('ERR_IPC_MESSAGE_SERIALIZE_ERROR');
+      }
+    } else {
+      messageType = INTERNAL_IPC_MESSAGE_TYPE.STRING;
+      if (typeof message !== 'string') {
+        if (message.toString) {
+          message = message.toString();
+        } else {
+          throw new Error('ERR_IPC_MESSAGE_TYPE_INVALID');
+        }
```

Comment:
Ditto, we could add a method `serialize()` on `IpcMessage` to make it read easier :)